### PR TITLE
[DNM] PR Test jwplayer/hls.js upgrade to v0.13.x

### PR DESCRIPTION
**Pull request by johnBartos**
### Checklist

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] API or design changes are documented in API.md

### CODE REVIEW

File: src/demux/adts.js [hunk 1]
```diff
@@ -228,8 +229,6 @@ export function appendFrame (track, data, offset, pts, frameIndex) {
     };
 
     track.samples.push(aacSample);
-    track.len += frameLength;
```
Comment (johnBartos) [inline src/demux/adts.js]: Not sure why we lost this

File: src/demux/id3.js [hunk 2]
```diff
@@ -294,6 +296,20 @@ class ID3 {
    * This library is free.  You can redistribute it and/or modify it.
    */
   static _utf8ArrayToStr (array, exitOnNull = false) {
+    const decoder = getTextDecoder();
```
Comment (johnBartos) [inline src/demux/id3.js]: I think there's an upstream PR that fixes an issue with this

File: src/demux/mpegaudio.js [hunk 3]
```diff
@@ -67,7 +67,6 @@ const MpegAudio = {
       track.channelCount = header.channelCount;
       track.samplerate = header.sampleRate;
       track.samples.push(sample);
-      track.len += header.frameLength;
```
Comment (johnBartos) [inline src/demux/mpegaudio.js]: Oh I remember - mangui deleted this for the track len exception bug

File: src/utils/webvtt-parser.js [hunk 4]
```diff
@@ -97,8 +98,10 @@ const WebVTTParser = {
         cueOffset = presentationTime - vttCCs.presentationOffset;
       }
 
-      cue.startTime += cueOffset - localTime;
-      cue.endTime += cueOffset - localTime;
+      if (timestampMap) {
```
Comment (johnBartos) [inline src/utils/webvtt-parser.js]: https://github.com/video-dev/hls.js/pull/#2179

File: src/hls.ts [hunk 5]
```diff
@@ -442,30 +463,50 @@ export default class Hls extends Observer {
...
-    hls.levelController.startLevel = newLevel;
+    this.levelController.startLevel = newLevel;
+  }
+
+  /**
+   * set  dynamically set capLevelToPlayerSize against (`CapLevelController`)
+   *
+   * @type {boolean}
+   */
+  set capLevelToPlayerSize (shouldStartCapping: boolean) {
```
Comment (johnBartos) [inline src/hls.ts]: https://github.com/video-dev/hls.js/pull/1986

File: src/controller/audio-stream-controller.js [hunk 6]
```diff
@@ -250,7 +250,9 @@ class AudioStreamController extends BaseStreamController {
             // we force a frag loading in audio switch as fragment tracker might not have evicted previous frags in case of quick audio switch
             this.fragCurrent = frag;
             if (audioSwitch || this.fragmentTracker.getState(frag) === FragmentState.NOT_LOADED) {
-              this.startFragRequested = true;
+              if (frag.sn !== 'initSegment') {
```
Comment (johnBartos) [inline src/controller/audio-stream-controller.js]: https://github.com/video-dev/hls.js/pull/2196

File: src/controller/stream-controller.js [hunk 7]
```diff
@@ -424,7 +424,9 @@ class StreamController extends BaseStreamController {
     let fragState = this.fragmentTracker.getState(frag);
 
     this.fragCurrent = frag;
-    this.startFragRequested = true;
+    if (frag.sn !== 'initSegment') {
```
Comment (johnBartos) [inline src/controller/stream-controller.js]: https://github.com/video-dev/hls.js/pull/2196

File: src/demux/id3.js [hunk 8]
```diff
@@ -239,7 +241,7 @@ class ID3 {
       [1-?] = {Description}\0{Value}
       */
       let index = 1;
-      const description = ID3._utf8ArrayToStr(frame.data.subarray(index));
+      const description = ID3._utf8ArrayToStr(frame.data.subarray(index), true);
```
Comment (johnBartos) [inline src/demux/id3.js]: https://github.com/video-dev/hls.js/pull/2217

File: src/is-supported.ts [hunk 9]
```diff
@@ -1,8 +1,11 @@
 import { getMediaSource } from './utils/mediasource-helper';
 
-export function isSupported () {
+export function isSupported (): boolean {
   const mediaSource = getMediaSource();
-  const sourceBuffer = window.SourceBuffer || window.WebKitSourceBuffer;
+  if (!mediaSource) {
```
Comment (johnBartos) [inline src/is-supported.ts]: https://github.com/video-dev/hls.js/pull/2305

File: src/controller/audio-stream-controller.js [hunk 10]
```diff
@@ -657,7 +685,9 @@ class AudioStreamController extends BaseStreamController {
         stats.tbuffered = performance.now();
         hls.trigger(Event.FRAG_BUFFERED, { stats: stats, frag: frag, id: 'audio' });
         let media = this.mediaBuffer ? this.mediaBuffer : this.media;
-        logger.log(`audio buffered : ${TimeRanges.toString(media.buffered)}`);
+        if (media) {
```
Comment (johnBartos) [inline src/controller/audio-stream-controller.js]: https://github.com/video-dev/hls.js/pull/2172

File: src/controller/audio-stream-controller.js [hunk 11]
```diff
@@ -423,7 +443,12 @@ class AudioStreamController extends BaseStreamController {
           logger.log(`start time offset found in playlist, adjust startPosition to ${startTimeOffset}`);
           this.startPosition = startTimeOffset;
         } else {
-          this.startPosition = 0;
+          if (newDetails.live) {
```
Comment (johnBartos) [inline src/controller/audio-stream-controller.js]: https://github.com/video-dev/hls.js/pull/2067

File: src/utils/logger.js [hunk 12]
```diff
@@ -49,8 +49,9 @@ function exportLoggerFunctions (debugConfig, ...functions) {
   });
 }
 
-export var enableLogs = function (debugConfig) {
-  if (debugConfig === true || typeof debugConfig === 'object') {
+export const enableLogs = function (debugConfig) {
+  // check that console is available
+  if ((global.console && debugConfig === true) || typeof debugConfig === 'object') {
```
Comment (johnBartos) [inline src/utils/logger.js]: https://github.com/video-dev/hls.js/pull/2095

File: src/remux/mp4-remuxer.js [hunk 13]
```diff
@@ -543,38 +542,51 @@ class MP4Remuxer {
       }
     }
 
+    // compute mdat size, as we eventually filtered/added some samples
+    let nbSamples = inputSamples.length;
+    let mdatSize = 0;
+    while (nbSamples--) {
```
Comment (johnBartos) [inline src/remux/mp4-remuxer.js]: https://github.com/video-dev/hls.js/pull/2096

File: src/remux/mp4-remuxer.js [hunk 14]
```diff
@@ -8,10 +8,12 @@ import MP4 from './mp4-generator';
 import Event from '../events';
 import { ErrorTypes, ErrorDetails } from '../errors';
 
+import { toMsFromMpegTsClock, toMpegTsClockFromTimescale, toTimescaleFromScale } from '../utils/timescale-conversion';
+
 import { logger } from '../utils/logger';
 
-// 10 seconds
-const MAX_SILENT_FRAME_DURATION = 10 * 1000;
+const MAX_SILENT_FRAME_DURATION_90KHZ = toMpegTsClockFromTimescale(10);
```
Comment (johnBartos) [inline src/remux/mp4-remuxer.js]: https://github.com/video-dev/hls.js/pull/2354

**Comment by johnBartos** [comment]
test this please

**Comment by robwalch** [comment]
test this please

**Pull request closed without merging**